File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        return search(nums,threshold);
    }
    
    int maxElem(int[] arr){
        int max = arr[0];
        for(int i = 1; i<arr.length; i++){
            if(max < arr[i]){
                max = arr[i];
            }
        }
        
        return max;
    }
    
    int sumOfDivisions(int[] arr, int diviser){
        int sum = 0;
        for(int i = 0; i<arr.length; i++){
            sum += Math.ceil((double)arr[i]/ (double)diviser);
        }
        
        return sum;
    }
    
    int search(int[] arr, int threshold){
        int start = 1;
        int end = maxElem(arr);
        
        while(start<=end){
            int mid = (start + end)/2;
            int result = sumOfDivisions(arr,mid);
            
            if(result<=threshold){
                end = mid-1;
            }else{
                start = mid+1;
            }
                
        }
        
        return start;
    }
}

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int maxElem(vector<int>& arr){
        int n = arr.size();
        int max = arr[0];
        for(int i = 1; i<n; i++){
            if(arr[i]>max) max = arr[i];
        }
        return max;
    }
    
    double foundDivisor(vector<int>& arr, int mid){
        int n = arr.size();
        double ans = 0;
        for(int i = 0; i<n; i++){
            ans += ceil((double) (arr[i]) / (double)(mid));
        }
        return ans;
    }
    
    int smallestDivisor(vector<int>& nums, int threshold) {
        int n = nums.size();
        if(n>threshold) return -1;
        int low = 0, high =  maxElem(nums);
        while(low<=high){
            int mid = (low+high)/2;
            double divisor = foundDivisor(nums,mid);
            if(divisor<=threshold){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return low;
    }
};